cluster-autoscaler/cloudprovider/aws/aws_util.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws/endpoints"
+	"io/ioutil"
+	"k8s.io/klog"
+	"net/http"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+var (
+	ec2MetaDataServiceUrl        = "http://169.254.169.254/latest/dynamic/instance-identity/document"
+	ec2PricingServiceUrlTemplate = "https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/%s/index.json"
+)
+
+type response struct {
+	Products map[string]product `json:"products"`
+}
+
+type product struct {
+	Attributes productAttributes `json:"attributes"`
+}
+
+type productAttributes struct {
+	InstanceType string `json:"instanceType"`
+	VCPU         string `json:"vcpu"`
+	Memory       string `json:"memory"`
+	GPU          string `json:"gpu"`
+}
+
+// GenerateEC2InstanceTypes returns a map of ec2 resources
+func GenerateEC2InstanceTypes(region string) (map[string]*InstanceType, error) {
+	instanceTypes := make(map[string]*InstanceType)
```

Comment:
No size hint?

Reply:
@seh  Not sure why I can not find your comments in this page.. I remember you left some comments here. Did you say it's better to have a initial capacity? The list is growing every quarter.

Reply:
Force pushes to files with comments hide them. I'm not sure why sometimes GitHub will retain them, but hide them in an "obsolete" block, or something to that effect.

In any case, I suggest you size the map to something like 200. Starting with the default size is going to induce reallocation and rehashing, given that we know there are more instance types than the default map size.

Reply:
I agree that go internally handles allocations and rehashing may hit performance. Not sure how golang implements it, probably exponential? rehashing for < 300 items may not be a problem. This method is only used during CA startup or the time we manually generate static list. Performance should not be a concern here.

The other reason I don't want to have a number here. If we set something like 200, it will be resized for some regions. Developers maybe also confused by the magic number there, this number is hard to tell from code itself.

Reply:
That's a fine reason. I just wanted to make sure that this is a choice, and not just a missed opportunity.